http/basic/src/test/java/io/cloudevents/http/HttpMessageReaderWriterTest.java

```diff
@@ -0,0 +1,247 @@
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Stream;
+
+import static io.cloudevents.core.test.Data.*;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class HttpMessageReaderWriterTest {
+
+    @ParameterizedTest
+    @MethodSource("io.cloudevents.core.test.Data#allEventsWithoutExtensions")
+    void testWriteStructured(CloudEvent event) {
+
+        final AtomicReference<byte[]> body = new AtomicReference<>();
+        final Map<String, String> headers = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+        GenericStructuredMessageReader.from(event, CSVFormat.INSTANCE).visit(
+            HttpMessageFactory.createWriter(headers::put, body::set)
+        );
+        assertThat(headers.get("content-type"))
+            .isEqualTo(CSVFormat.INSTANCE.serializedContentType());
+        assertThat(body.get())
+            .isEqualTo(CSVFormat.INSTANCE.serialize(event));
+
+    }
+
+    @ParameterizedTest
+    @MethodSource("binaryTestArguments")
+    void testWriteBinary(CloudEvent event, Map<String,String> expectedHeaders, byte[] expectedBody) {
+
+        final AtomicReference<byte[]> body = new AtomicReference<>();
+        final Map<String, String> headers = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+
+        new HttpMessageWriter(headers::put, body::set).writeBinary(event);
+
+        expectedHeaders.forEach((k, v) -> {
+            assertThat(headers.get(k)).isEqualTo(v);
+        });
+        if (expectedBody != null) {
+            assertThat(body.get()).isEqualTo(expectedBody);
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("io.cloudevents.core.test.Data#allEventsWithoutExtensions")
+    public void testReadStructured(CloudEvent event) {
+        byte[] serializedEvent = CSVFormat.INSTANCE.serialize(event);
+
+        Map<String,String> headers = new HashMap<String,String>() {{
+            put("content-type", CSVFormat.INSTANCE.serializedContentType() + "; charset=utf8");
+        }};
+
+        MessageReader message = HttpMessageFactory.createReader(
+            headers,
+            serializedEvent
+        );
+
+        assertThat(message.getEncoding())
+            .isEqualTo(Encoding.STRUCTURED);
+        assertThat(message.toEvent())
+            .isEqualTo(event);
+    }
+
+    @ParameterizedTest
+    @MethodSource("binaryTestArguments")
+    public void testReadBinary(CloudEvent expectedEvent, Map<String,String> headers, byte[] body) {
+        MessageReader message = HttpMessageFactory.createReader(headers::forEach, body);
+
+        assertThat(message.getEncoding())
+            .isEqualTo(Encoding.BINARY);
+        assertThat(message.toEvent())
+            .isEqualTo(expectedEvent);
+    }
+
+    public static Stream<Arguments> binaryTestArguments() {
+        return Stream.of(
+            // V03
+            Arguments.of(
+                V03_MIN,
+                new HashMap<String,String>() {{
```

Comment:
[![Every time you use double-brace initialization, God kills a kitten](https://lukaseder.files.wordpress.com/2014/12/58731480.jpg)](https://blog.jooq.org/2014/12/08/dont-be-clever-the-double-curly-braces-anti-pattern/)

Probably OK in a test, but you could also write a helper function that takes a `String[]` or a `String[][]` and makes a `Map` out of it. Better still would be to use `Map.of(...)`, but that requires Java 9+.

Reply:
Originally I used something different: https://github.com/cloudevents/sdk-java/pull/225/commits/c09ef8c7f400a2ce97ef1de19f42f06ec4ab120c#diff-da38f4f0dad4b560cd509239f96a762eL241 was it better?

Reply:
I didn't realize it uses anon classes. Gee couldn't they just already add proper literals to the language.

Reply:
Of course this is only a test, so the kittens God is killing are only test kittens. Still, I think I preferred the way you had it before, which was at least as readable without triggering the kitten-killing fear in seasoned Java developers.

For literals, I think `List.of(...)` and `Set.of(...)`, added in Java 9, are good enough for those types without needing a special syntax. There's also `Map.of(...)`, which you could have used here except for the desire to work on Java 8. But that one is a bit dicey anyway since there's no obvious syntactic difference between keys and values, and it arbitrarily stops at 10 key-value pairs.